sns/parity_math.c

```diff
@@ -385,85 +768,911 @@ static void xor_calculate(struct m0_parity_math *math,
+		pe = 0;
+		for (ui = 0; ui < math->pmi_data_count; ++ui)
+			if (ui != failure_index)
+				pe ^= (m0_parity_elem_t)((uint8_t*)
+				       data[ui].b_addr)[ei];
+
+		if (ui != failure_index)
+			((uint8_t*)data[failure_index].b_addr)[ei] = pe ^
+				((uint8_t*)parity[0].b_addr)[ei];
+		else /* Parity was lost, so recover it. */
+			((uint8_t*)parity[0].b_addr)[ei] = pe;
 	}
 }
 
-static void reed_solomon_encode(struct m0_parity_math *math,
-				const struct m0_buf *data,
-				struct m0_buf *parity)
+/** @todo Iterative reed-solomon decode to be implemented. */
+static void fail_idx_reed_solomon_recover(struct m0_parity_math *math,
+					  struct m0_buf *data,
+					  struct m0_buf *parity,
+					  const uint32_t failure_index)
 {
-#define PARITY_MATH_REGION_ENABLE 0
+}
 
-#if !PARITY_MATH_REGION_ENABLE
-	uint32_t	  ei; /* block element index. */
-#endif
-	uint32_t	  pi; /* parity unit index. */
-	uint32_t	  di; /* data unit index. */
-	m0_parity_elem_t  mat_elem;
-	uint32_t	  block_size = data[0].b_nob;
+static int ir_si_blocks_init(struct m0_sns_ir *ir)
+{
+	uint32_t i;
+	uint32_t blocks_count;
+	int      ret = 0;
+
+	M0_PRE(ir != NULL);
+
+	blocks_count = ir_blocks_count(ir);
+	M0_ALLOC_ARR(ir->si_blocks, blocks_count);
+	if (ir->si_blocks == NULL)
+		return BUF_ALLOC_ERR_INFO(-ENOMEM, "ir blocks", blocks_count);
+
+	for (i = 0; i < blocks_count; ++i) {
+		ir->si_blocks[i].sib_idx = i;
+		ir->si_blocks[i].sib_status = M0_SI_BLOCK_ALIVE;
+		ret = m0_bitmap_init(&ir->si_blocks[i].sib_bitmap,
+				     blocks_count);
+		if (ret != 0)
+			return M0_ERR_INFO(ret, "Failed to initialize bitmap "
+					  "for si_blocks with index = %u", i);
+	}
+
+	return ret;
+}
+
+static inline uint32_t ir_blocks_count(const struct m0_sns_ir *ir)
+{
+	return ir->si_data_nr + ir->si_parity_nr;
+}
+
+static void gfaxpy(struct m0_bufvec *y, struct m0_bufvec *x,
+		   m0_parity_elem_t alpha)
+{
+	uint32_t                i;
+	uint32_t                seg_size;
+	uint8_t                *y_addr;
+	uint8_t                *x_addr;
+	m0_bcount_t             step;
+	struct m0_bufvec_cursor x_cursor;
+	struct m0_bufvec_cursor y_cursor;
+
+	M0_ENTRY("y=%p, x=%p, alpha=%u", y, x, (uint32_t)alpha);
+
+	M0_PRE(y != NULL && x != NULL);
+	M0_PRE(y->ov_vec.v_nr != 0);
+	M0_PRE(y->ov_vec.v_nr == x->ov_vec.v_nr);
+	M0_PRE(y->ov_vec.v_count[0] != 0);
+	M0_PRE(y->ov_vec.v_count[0] == x->ov_vec.v_count[0]);
+	seg_size = y->ov_vec.v_count[0];
+	M0_PRE(m0_forall(i, y->ov_vec.v_nr, y->ov_vec.v_count[i] == seg_size));
+
+	m0_bufvec_cursor_init(&y_cursor, y);
+	m0_bufvec_cursor_init(&x_cursor, x);
+	do {
+		x_addr  = m0_bufvec_cursor_addr(&x_cursor);
+		y_addr  = m0_bufvec_cursor_addr(&y_cursor);
+
+		switch (alpha) {
+		/* This is a special case of the 'default' case that follows.
+		 * Here we avoid unnecessary call to gmul.*/
+		case 1:
+			for (i = 0; i <seg_size; ++i) {
+				y_addr[i] = gadd(y_addr[i],
+						 x_addr[i]);
+			}
+			break;
+		default:
+			for (i = 0; i < seg_size; ++i) {
```

Comment:
[Style] Remove the open braces for a case of single statement in a for loop.

Reply:
Done.